Create single inso executable using pkg

develohpanda opened the pull request:
### GitHub Actions workflow changes

The release workflow has been updated to also create a GitHub release and publish Inso CLI binary artifacts into the release. The Insomnia app release is created against the `core@2021.2.3` tag, while the Inso CLI release is created agsinst the `lib@1.2.3` tag.

This is because Inso CLI is also published to NPM, using the `lib@1.2.3` tag generated from Lerna, and this way both distributions (what's available in NPM and what's available in GitHub Releases) for Inso CLI remain in sync.

The **Release** workflow will now:
* Run all tests (excluding smoke tests)
* Publish to NPM
* Assert if a release already exists against a `lib@x.y.z` tag.
	* Exists: skip creating and publishing Inso CLI binaries
	* Doesn't exist:
		* Create release
		* Create Inso CLI binaries on macOS, Windows and Ubuntu
		* Run CLI smoke tests
		* Publish the compressed binaries into the GitHub Release
			* `windows.zip`
			* `macos.tar`
			* `linux.tar.xz`
* Build Insomnia app for macOS, Windows, Ubuntu
* Publish Insomnia app artifacts into GitHub Release

The **Test/PR** workflow will now:
* Create Inso CLI binaries for macOS, Windows and Ubuntu
* Publish the compressed binaries into the CI artifacts (retained for ~5 days)
* Run CLI smoke tests

### Code changes

* Plugins loading now has a try-catch. In case a plugin fails to load, it will not crash anything and will print to console.
* `jsonpath` dependency updated to `jsonpath-plus`
* Moved `deep-equal` to a `devDependency`

### Smoke tests

* Updated `inso-nedb` fixture
* Added CLI smoke tests for more commands (`inso run test`, `inso generate config`, `inso export spec`, `inso lint spec`)
* CLI smoke tests also run with the compiled binary

### Plugins

* Updated `insomnia-plugin-jsonpath` to use `jsonpath-plus` instead of `jsonpath`
* Updated `insomnia-plugin-os` to use `jsonpath-plus` instead of `jsonpath`
* Updated `insomnia-plugin-response` to use `jsonpath-plus` instead of `jsonpath` (this impacts request chaining)

### Local dev

```
npm run clean
npm run bootstrap
npm run inso-package
npm run test:smoke:cli
```

develohpanda committed:
package inso binaries from code

develohpanda committed:
add inso package script to root

develohpanda commented on package.json:
```diff
@@ -77,4 +78,4 @@
     "typescript": "^4.2.3"
   }
-}
+}
```
```suggestion
}

```

develohpanda committed:
Update package.json

develohpanda commented on packages/insomnia-app/package.json:
```diff
@@ -225,6 +224,7 @@
     "concurrently": "^3.5.0",
     "cross-env": "^2.0.0",
     "css-loader": "^2.1.1",
+    "deep-equal": "^1.0.1",
```
Moved from `dependencies` to `devDependencies`. If we don't do this, after running through `pkg` we face this deeply nested error

![image](https://user-images.githubusercontent.com/4312346/134669604-a9fe98ef-6d74-4fc1-a678-1154c03154ae.png)

This error likely arises from `deep-equal` as indicated in https://github.com/vercel/pkg/issues/943. I moved it to a dev dependency in order to create a shim for send-request (like we do for electron), but a shim wasn't necessary and simply moving it to a `devDependency` seemed to do the trick.

wdawson replied:
Hopefully, this doesn't break places in the app where we're using this?

develohpanda replied:
Shouldn't have any material impact because we use webpack. It just changes how webpack handles it (but we have limited use of it and can/should double check)

develohpanda replied:
Tried to revert this change, but it is necessary, otherwise the error shows again.

develohpanda commented on packages/insomnia-inso/package.json:
```diff
@@ -85,4 +93,4 @@
     "yaml": "^1.10.0"
   }
-}
+}
```
```suggestion
}

```

develohpanda committed:
Update packages/insomnia-inso/package.json

dimitropoulos commented on packages/insomnia-inso/package.json:
```diff
@@ -31,7 +31,14 @@
+    "prepare": "npm run build:production",
+    "package": "rimraf binaries && npm run build:production && pkg ."
+  },
+  "pkg": {
```
looks like we can point to a specific config file like `pkg . --config pkg.config.json` which could be better than the implicit package.json approach (since, it is more resilient to pwd differences).

dimitropoulos replied:
nix that.  this didn't work after 30 seconds of me trying it and therefore is probably not worth the effort.  I saw https://github.com/vercel/pkg/issues/1302 and then was all like
![perrie-edwards-jesy-nelson](https://user-images.githubusercontent.com/15232461/134713065-8336e439-033e-4b96-92d8-c0d2e1c86b29.gif)

dimitropoulos committed:
adds failing test for simultaneously smoke testing the single executable

also fixes some other things, e.g. using the (now deprecated) `--appDataDir` option.

also adds a (hopefully) helpful message for if this breaks in the future because of the files moving

dimitropoulos committed:
adds verify-pkg script and cross-compile targets

develohpanda committed:
bootstrap and update packagelock

develohpanda committed:
WIP

dimitropoulos commented:
in case it's useful, here's where it's at for me right now:

<details>
<summary>npm run test:cli
</summary>

```
dimitri@dimitri:~/src/github.com/Kong/insomnia/packages/insomnia-smoke-test$ npm run test:cli

> insomnia-smoke-test@2.3.2 test:cli /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-smoke-test
> npm run with-mock "npm run cli"


> insomnia-smoke-test@2.3.2 with-mock /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-smoke-test
> concurrently --names server,app --success first --kill-others "npm run serve" "npm run cli"

[app] 
[app] > insomnia-smoke-test@2.3.2 cli /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-smoke-test
[app] > jest --detectOpenHandles --testPathPattern cli
[app] 
[server] 
[server] > insomnia-smoke-test@2.3.2 serve /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-smoke-test
[server] > ts-node server/index.ts
[server] 
[app] watchman warning:  Recrawled this watch 1 times, most recently because:
[app] dir missing from internal stateTo resolve, please review the information on
[app] https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
[app] To clear this warning, run:
[app] `watchman watch-del /home/dimitri/src/github.com/Kong/insomnia ; watchman watch-project /home/dimitri/src/github.com/Kong/insomnia`
[app] 
[server] Listening at http://localhost:4010
[app]  FAIL  cli/app.test.ts (53.187 s)
[app]   ● run test › can run unit test with /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/bin/inso
[app] 
[app]     Command failed with exit code 1: /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/bin/inso run --src fixtures/inso-nedb test --env Dev TestSuite
[app] 
[app]     
[app]       TestSuite
[app]     [network] Sending req=req_5a77fa051cae441783d6e20a364a60f7 env=env_baf32bd57d0e45f38e3914e51d2f0c29
[app]     delay millis: 953
[app]     fetching stuff
[app]     fetched requests
[app]     fetched settings
[app]     fetched ancestors
[app]     fetched environment
[app]     get rendered request and context
[app]     fetched cookie jar
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     render string {
[app]       x: 'http://localhost:4010',
[app]       context: {
[app]         getMeta: [Function],
[app]         getKeysContext: [Function],
[app]         getPurpose: [Function],
[app]         getExtraInfo: [Function],
[app]         getEnvironmentId: [Function],
[app]         getProjectId: [Function],
[app]         base_url: 'http://localhost:4010'
[app]       },
[app]       path: 'Environment'
[app]     }
[app]     5
[app]     render
[app]     description
[app]     set default protocol
[app]     fetched get rendered request
[app]     fetched workspace
[app]     applying plugin hooks
[app]     [network] actually sending
[app]     [network] init curl
[app]     [network] curl.perform
[app]     [network] Response failed req=req_5a77fa051cae441783d6e20a364a60f7 err=Error: Couldn't resolve host name
[app]         1) Returns 200
[app] 
[app] 
[app]       0 passing (1s)
[app]       1 failing
[app] 
[app]       1) TestSuite
[app]            Returns 200:
[app]          AssertionError: expected undefined to equal 200
[app] 
[app]       at Context.<anonymous> (../../../../../../../../tmp/insomnia-testing/0.8843776332476532-test.ts:11:32)
[app]       
[app]       at makeError (node_modules/execa/lib/error.js:59:11)
[app]       at Function.Object.<anonymous>.module.exports.sync (node_modules/execa/index.js:186:17)
[app]       at cli/app.test.ts:50:30
[app] 
[app]   ● run test › can run unit test with /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/binaries/insomnia-inso-linux
[app] 
[app]     Command failed with exit code 1: /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/binaries/insomnia-inso-linux run --src fixtures/inso-nedb test --env Dev TestSuite
[app] 
[app]     
[app]       TestSuite
[app]     [network] Sending req=req_5a77fa051cae441783d6e20a364a60f7 env=env_baf32bd57d0e45f38e3914e51d2f0c29
[app]     delay millis: 970
[app]     fetching stuff
[app]     fetched requests
[app]     fetched settings
[app]     fetched ancestors
[app]     fetched environment
[app]     get rendered request and context
[app]     fetched cookie jar
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     start next
[app]     render string {
[app]       x: 'http://localhost:4010',
[app]       context: {
[app]         getMeta: [Function],
[app]         getKeysContext: [Function],
[app]         getPurpose: [Function],
[app]         getExtraInfo: [Function],
[app]         getEnvironmentId: [Function],
[app]         getProjectId: [Function],
[app]         base_url: 'http://localhost:4010'
[app]       },
[app]       path: 'Environment'
[app]     }
[app]         1) Returns 200
[app] 
[app] 
[app]       0 passing (50s)
[app]       1 failing
[app] 
[app]       1) TestSuite
[app]            Returns 200:
[app]          Error: Timeout of 50000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/tmp/insomnia-testing/0.039154759921680826-test.ts)
[app] 
[app]       48 |     }
[app]       49 | 
[app]     > 50 |     const { failed } = execa.sync(binPath, [
[app]          |                              ^
[app]       51 |       'run',
[app]       52 |       '--src', 'fixtures/inso-nedb',
[app]       53 |       'test',
[app] 
[app]       
[app]       at makeError (node_modules/execa/lib/error.js:59:11)
[app]       at Function.Object.<anonymous>.module.exports.sync (node_modules/execa/index.js:186:17)
[app]       at cli/app.test.ts:50:30
[app] 
[app] Test Suites: 1 failed, 1 total
[app] Tests:       2 failed, 2 total
[app] Snapshots:   0 total
[app] Time:        53.224 s
```
</details>

dimitropoulos committed:
adds back cross compilation and check script

develohpanda committed:
remove console logs

develohpanda committed:
undo more things

develohpanda committed:
add safety-net around plugin loading

develohpanda committed:
convert insomnia-plugin-jsonpath to ts and webpack in an attempt to overcome pkg failing to find insomnia-plugin-jsonpath > jsonpath > esprima

develohpanda committed:
replace all jsonpath with jsonpath-plus

develohpanda committed:
remove debug script

develohpanda committed:
script to override pkg targets

develohpanda committed:
add CLI package step to test workflow

develohpanda committed:
fix test

develohpanda committed:
only target node12-x64 for the OS being run on

develohpanda committed:
upload binary artifacts for OS

develohpanda committed:
extract TS move to separate PR

develohpanda committed:
refactor cli smoke test and add ramda

develohpanda committed:
write oa3 spec for smoke test server, update inso-nedb fixture to include a spec so we can lint and generate config with it, add readme for how to update it

develohpanda committed:
add smoke tests for other commands and move docs to main readme

develohpanda committed:
remove libc detection

develohpanda committed:
update test workflow to tar and upload

develohpanda committed:
improve tar structure and package from script

develohpanda committed:
add test case for export spec

develohpanda committed:
YAMLING

develohpanda committed:
update packagelock and tar command

develohpanda commented on packages/insomnia-inso/src/scripts/pkg.ts:
```diff
@@ -0,0 +1,74 @@
+  if (isMac()) {
+    return ['node12-macos-x64'];
+  } else if (isLinux()) {
+    return ['node12-linux-x64'];
```
Should this include an alpine target as well?

dimitropoulos committed:
updates package-lock.json of insomnia-plugin-response

(all I did was run a bootstrap, so I'm assuming this is correct)

dimitropoulos committed:
fixes error handling (and removes some else-returns) in pkg.ts

prior to this, the `reject` on line 28 does not hault execution.  it's reasonable that this mistake was made because it's an easy one to make if you don't allow the function that ultimately had the failure case to have the control flow.

As it was before, code _after_ the reject was still executing.

Now, the promise rejection is handled in the place it really occurs.  Promises are designed to work well with `throw` like this and since there is no other callback that the throw is captured in (e.g. like a setTimeout or something) this is totally fine to do.  Here is what the output would look like:

```
> ts-node src/scripts/pkg.ts

Error: [pkg-inso] Unsupported OS: freebsd
    at getTargets (/home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/src/scripts/pkg.ts:24:9)
    at /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/src/scripts/pkg.ts:29:21
    at new Promise (<anonymous>)
    at pkg (/home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/src/scripts/pkg.ts:28:10)
    at Object.<anonymous> (/home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/src/scripts/pkg.ts:67:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Module.m._compile (/home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/node_modules/ts-node/src/index.ts:1056:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Object.require.extensions.<computed> [as .ts] (/home/dimitri/src/github.com/Kong/insomnia/packages/insomnia-inso/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (internal/modules/cjs/loader.js:985:32)
```

dimitropoulos committed:
simplifies platform getting

there's no way for the platform to change in the midst of a script like this, and therefore no motivation to have a closure for it.

dimitropoulos committed:
adds documentation and simplifies targets

this was returning an array before in part because it needed to check if the array was empty to detect an error case.  now, it doesn't need to do that, so there's no point in returning an array that statically only has one item just to join it with commas later (which will have no effect since it only has one item).  If we ever want to modify this, we can just as well change something like `'node12-macos-x64'` to `'node12-macos-x64,node12-macos-arm64'` in the `getTargets` function.

dimitropoulos committed:
fixes typo now that there's only one binary

before, this made sense to be plural because it was always expected that there'd be more than one binary

develohpanda committed:
compress via script

develohpanda committed:
update linux compression command to use xz compressino

develohpanda committed:
Update compression args

develohpanda committed:
release action WIP

develohpanda commented on .github/workflows/release-core.yml:
```diff
@@ -53,8 +53,61 @@ jobs:
+      - name: Get tag name for version
+        id: get-tag-name
+        run: echo "::set-output tag-name=lib@${{ steps.package-version.outputs.current-version }}"
+      - name: Create GitHub Release
```
Only create release if release _doesn't_ exist for the package.json version (meaning it's a new version of inso being published)

develohpanda committed:
update release script to only create a release if it doesnt exist

develohpanda committed:
fix step output

develohpanda committed:
handle error when no release is found for tag

develohpanda committed:
fix condition

develohpanda committed:
debug steps

develohpanda committed:
disable test workflow for now

develohpanda committed:
rename

develohpanda committed:
flip condition??

develohpanda committed:
should skip because the release already exists

develohpanda committed:
should skip

develohpanda committed:
should skip

develohpanda committed:
should skip

develohpanda committed:
BINGO! This will skip

develohpanda committed:
this should create a release because I deleted the existing one

develohpanda committed:
try json conversion

develohpanda committed:
check as string

develohpanda committed:
skip tests for now

develohpanda committed:
switch test workflow on

develohpanda committed:
does moving the npm target env variables fix node-libcurl for inso?

develohpanda committed:
uncomment release workflow changes

develohpanda committed:
fix typo in release workflow

johnwchadwick reviewed (approved):
🚀

dimitropoulos reviewed (approved):
(on linux)

I generated specs, linted specs, ran tests, passed local files, exported specs, saw help output.

all worked exactly as expected!

develohpanda committed:
Merge branch 'develop' into inso-binaries
